### ISSUE TITLE
v2.0.0 Ideas

### ISSUE THREAD

**Issue by yohamta** [yohamta/dagu#381]
I'm looking to make Dagu more useful and practical for a wider range of use cases. Here are some ideas I have in mind for the next major version:

- [ ] Implement a robust file database to store execution history data, replacing the current JSON file system.
- [ ] Allow concurrent execution of DAGs, enabling users to run multiple instances at the same time.
- [ ] Add user management capabilities, giving users more control over authorization.
- [ ] Redesign the Web UI for a more intuitive user experience.
- [ ] Make it possible to use dagu as a Go module in other programs.
- [ ] Redesign the APIs using the OpenAPI specification, with an embeddable client library and documentation generated using Swagger.
- [ ] Support multiple workers, allowing users to manage multiple servers in a single web server instance.
- [ ] Support for Slack integration.
- [ ] Support for cross-DAG dependencies to allow for more complex workflows.

**Comment by spreadsheetfun** [comment]
Any idea about storing the jobs/runs in a database like postgres...so that its more scalable and also reliable/fault tolerant.

**Comment by yohamta** [comment]
Hey @spreadsheetfun 👋 
Thank you for the comment! I would like to keep it self-contained so that users don't need to install database separately. However, I think we can offer options for users to connect to a standalone Postgres database if they prefer.

**Comment by Arvintian** [comment]
About support integration with FaaS like aws Lambda?

Now, i use the http executor to do this.

**Comment by Arvintian** [comment]
About support integration with FaaS like aws Lambda?

Now, i use the http executor to do this.

**Comment by yohamta** [comment]
@Arvintian That sounds useful, let's explore the way to handle it.

**Comment by stuartskelton** [comment]
Could there be an option to say `after this task  wait  X seconds  before executing it the next  step?

**Comment by yohamta** [comment]
@stuartskelton Yes, currently, the `delaySec` option is available for DAG configuration. This can also be added to Step configuration, even for v1, such as `delaySecAfterFinish`.

```
delaySec: 1                          # Interval seconds between steps
```

**Comment by dispensable** [comment]
Is it possible to add a Prometheus like metrics system to expose dags status ? Those metrics can be used to monitoring SLA miss or critical event during dag running.

**Comment by yohamta** [comment]
@dispensable Sounds nice. Would it be possible to see an example OSS that has a feature similar to it? Maybe we can look at it to consider our implementation.

**Comment by dispensable** [comment]
https://github.com/robinhood/airflow-prometheus-exporter

I haven't see single instance cron alternative combined with metrics. But there are many airflow prometheus exporter.

**Comment by sidneiweber** [comment]
Slack integration is interesting

**Comment by shbfy** [comment]
DAGs running other DAGs would be neat!

**Comment by yohamta0** [comment]
This omnibus feature-ideas issue is outdated now. Several items have since been implemented or have more focused issues, and the remaining ideas are easier to track individually.

Closing this tracker to keep the backlog actionable. Please open separate focused issues for any remaining feature ideas that are still relevant.
